src/dasi/core/AggregatedReadHandle.cc

```diff
@@ -0,0 +1,63 @@
+
+#include "dasi/core/AggregatedReadHandle.h"
+
+#include "dasi/util/Exceptions.h"
+
+#include <iostream>
+#include <ostream>
+
+namespace dasi::core {
+
+AggregatedReadHandle::AggregatedReadHandle(const std::vector<ReadHandle*>& handles) :
+    handles_(handles.begin(), handles.end()),
+    current_(handles_.begin()) {}
+
+AggregatedReadHandle::~AggregatedReadHandle() {
+    if (!handles_.empty()) {
+        std::cerr << "ERROR: " << *this << " has not been closed" << std::endl;
+    }
+}
+
+size_t AggregatedReadHandle::read(void* buf, size_t len) {
+
+    if (current_ == handles_.end()) {
+        return 0;
+    }
+
+    size_t nread = 0;
+    auto cur = static_cast<char*>(buf);
+    
+    do {
+        size_t chunk = (*current_)->read(static_cast<void*>(cur), len - nread);
+        cur += chunk;
+        nread += chunk;
+
+        if (nread < len) {
+            ++current_;
+            if (current_ == handles_.end()) {
+                return nread;
+            }
+            open();
+        }
+    } while (nread < len);
+
+    return nread;
+}
+
+void AggregatedReadHandle::open() {
+    (*current_)->open();
```

Comment:
You carefully ensure that the handles are opened lazily, and sequentially. But closing the handles is deferred until the end.

Either (a) open everything at once, or (b) close things as you go along.

I would take a look at the PooledHandle class in eckit while you ponder this. It is very likely that you are going to be reading a number of fields from a small number of files (potentially alternating between them). We will likely need a similar mechanism here.

Reply:
A second thought - if you are opening lazily, you probably want to use an AutoCloser which gets assigned in the open() function. When you open the next one, you reassign the AutoCloser - which results in the first file being closed. That way you don't have to have any explicit loop over h_->close() in the descructor - it just happens.

Note in a secondary sense, I think you will need an openInternal() or similar function. The externally accessible open() should open the readHandle() at the start, and give an error if you try and open a handle that is open and in the middle of being read...

Reply:
I think option b is probably the cleanest. If we want to implement a pooled ReadHandle, it will have to be tied to the specific type of handle anyway, so there is no point in trying to keep handles open in AggregatedReadHandle.

Reply:
Do that then :-).